src/utils/registry.ts

```diff
@@ -0,0 +1,272 @@
+    return os.arch() === 'x64' ? 'win64' : 'win32';
+  return platform as BrowserPlatform;
+})();
+
+export const registryDirectory = (() => {
+  const envDefined = getFromENV('PLAYWRIGHT_BROWSERS_PATH');
+  if (envDefined === '0')
+    return path.join('..', '..', '.local-browsers');
+  if (envDefined)
+    return path.resolve(process.cwd(), envDefined);
+  return path.join(cacheDirectory(), 'ms-playwright');
+})();
+
+export function isBrowserDirectory(browserDirectory: string): boolean {
+  const baseName = path.basename(browserDirectory);
+  for (const browserName of allBrowserNames) {
+    if (baseName.startsWith(browserName + '-'))
+      return true;
+  }
+  return false;
+}
+
+export class Registry {
+  private _descriptors: BrowserDescriptor[];
+
+  constructor(packagePath: string) {
+    const browsersJSON = JSON.parse(fs.readFileSync(path.join(packagePath, 'browsers.json'), 'utf8'));
+    this._descriptors = browsersJSON['browsers'];
+  }
+
+  _browser(browserName: BrowserName): BrowserDescriptor {
+    const browser = this._descriptors.find(browser => browser.name === browserName);
+    if (!browser)
+      throw new Error('Unknown browser name: ' + browserName);
+    return browser;
+  }
+
+  browserDirectory(browserName: BrowserName): string {
+    const browser = this._browser(browserName);
+    return path.join(registryDirectory, `${browser.name}-${browser.revision}`);
+  }
+
+  revision(browserName: BrowserName): string {
+    return this._browser(browserName).revision;
+  }
+
+  linuxLddDirectories(browserName: BrowserName): string[] {
+    const browserDirectory = this.browserDirectory(browserName);
+    if (browserName === 'chromium')
+      return [path.join(browserDirectory, 'chrome-linux')];
+    if (browserName === 'firefox')
+      return [path.join(browserDirectory, 'firefox')];
+    if (browserName === 'webkit') {
+      return [
+        path.join(browserDirectory, 'minibrowser-gtk'),
+        path.join(browserDirectory, 'minibrowser-gtk', 'bin'),
+        path.join(browserDirectory, 'minibrowser-gtk', 'lib'),
+        path.join(browserDirectory, 'minibrowser-wpe'),
+        path.join(browserDirectory, 'minibrowser-wpe', 'bin'),
+        path.join(browserDirectory, 'minibrowser-wpe', 'lib'),
+      ];
+    }
+    return [];
+  }
+
+  windowsExeAndDllDirectories(browserName: BrowserName): string[] {
+    const browserDirectory = this.browserDirectory(browserName);
+    if (browserName === 'chromium')
+      return [path.join(browserDirectory, 'chrome-win')];
+    if (browserName === 'firefox')
+      return [path.join(browserDirectory, 'firefox')];
+    if (browserName === 'webkit')
+      return [browserDirectory];
+    return [];
+  }
+
+  executablePath(browserName: BrowserName): string | undefined {
+    const browserDirectory = this.browserDirectory(browserName);
+    const tokens = EXECUTABLE_PATHS[browserName][hostPlatform];
+    return tokens ? path.join(browserDirectory, ...tokens) : undefined;
```

Comment:
Why do we support undefined here? Is there a usecase?

Reply:
playwright-core might return undefined